File: Payload_Type/sage/container/commands/list.go
package commands

import (
	// Standard
	"fmt"
	"strings"

	// Internal
	b "github.com/MythicAgents/sage/Payload_Type/sage/container/pkg/bedrock"
	"github.com/MythicAgents/sage/Payload_Type/sage/container/pkg/env"
	"github.com/MythicAgents/sage/Payload_Type/sage/container/pkg/ollama"
	"github.com/MythicAgents/sage/Payload_Type/sage/container/pkg/openai"
	"github.com/MythicAgents/sage/Payload_Type/sage/container/pkg/openwebui"

	// Mythic

	structs "github.com/MythicMeta/MythicContainer/agent_structs"
	"github.com/MythicMeta/MythicContainer/logging"
	"github.com/MythicMeta/MythicContainer/mythicrpc"
)

func list() structs.Command {
	attr := structs.CommandAttribute{
		SupportedOS: []string{"sage"},
	}

	provider := structs.CommandParameter{
		Name:                                    "provider",
		ModalDisplayName:                        "Provider",
		CLIName:                                 "provider",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_CHOOSE_ONE,
		Description:                             "The model provider you want to list models from",
		Choices:                                 append([]string{""}, env.ProvidersString()...),
		DefaultValue:                            "",
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       0,
				AdditionalInformation: nil,
			},
		},
	}

	apiEndpoint := structs.CommandParameter{
		Name:             "API_ENDPOINT",
		ModalDisplayName: "API Endpoint",
		CLIName:          "API-ENDPOINT",
		ParameterType:    structs.COMMAND_PARAMETER_TYPE_STRING,
		DefaultValue:     "",
		Description:      "[OPTIONAL] The API endpoint to use for the selected provider",
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       1,
				AdditionalInformation: nil,
			},
		},
	}

	apiKey := structs.CommandParameter{
		Name:             "API_KEY",
		ModalDisplayName: "API Key",
		CLIName:          "API-KEY",
		ParameterType:    structs.COMMAND_PARAMETER_TYPE_STRING,
		DefaultValue:     "",
		Description:      "[OPTIONAL] The API key to use for the selected provider",
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       2,
				AdditionalInformation: nil,
			},
		},
	}

	awsAccessKey := structs.CommandParameter{
		Name:             "AWS_ACCESS_KEY_ID",
		ModalDisplayName: "AWS Access Key ID",
		CLIName:          "AWS-ACCESS-KEY-ID",
		ParameterType:    structs.COMMAND_PARAMETER_TYPE_STRING,
		DefaultValue:     "",
		Description:      "[OPTIONAL] The AWS Access Key ID (AWS_ACCESS_KEY_ID) to use for Bedrock",
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       3,
				AdditionalInformation: nil,
			},
		},
	}

	awsSecretAccessKey := structs.CommandParameter{
		Name:             "AWS_SECRET_ACCESS_KEY",
		ModalDisplayName: "AWS Secret Access Key",
		CLIName:          "AWS-SECRET-ACCESS-KEY",
		ParameterType:    structs.COMMAND_PARAMETER_TYPE_STRING,
		DefaultValue:     "",
		Description:      "[OPTIONAL] The AWS Secret Access Key (AWS_SECRET_ACCESS_KEY) to use for Bedrock",
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       4,
				AdditionalInformation: nil,
			},
		},
	}

	awsSessionToken := structs.CommandParameter{
		Name:             "AWS_SESSION_TOKEN",
		ModalDisplayName: "AWS Session Token",
		CLIName:          "AWS-SESSION-TOKEN",
		ParameterType:    structs.COMMAND_PARAMETER_TYPE_STRING,
		DefaultValue:     "",
		Description:      "[OPTIONAL] The AWS Session Token (AWS_SESSION_TOKEN) to use for Bedrock",
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       5,
				AdditionalInformation: nil,
			},
		},
	}

	awsRegion := structs.CommandParameter{
		Name:             "AWS_DEFAULT_REGION",
		ModalDisplayName: "AWS Default Region",
		CLIName:          "AWS-DEFAULT-REGION",
		ParameterType:    structs.COMMAND_PARAMETER_TYPE_STRING,
		DefaultValue:     "",
		Description:      "[OPTIONAL] The AWS Region (AWS_DEFAULT_REGION) to use for Bedrock",
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       6,
				AdditionalInformation: nil,
			},
		},
	}

	command := structs.Command{
		Name:                           "list",
		NeedsAdminPermissions:          false,
		HelpString:                     "list",
		Description:                    "List all available from the selected provider",
		Version:                        0,
		SupportedUIFeatures:            nil,
		Author:                         "@Ne0nd0g",
		MitreAttackMappings:            []string{},
		ScriptOnlyCommand:              false,
		CommandAttributes:              attr,
		CommandParameters:              []structs.CommandParameter{provider, apiEndpoint, apiKey, awsAccessKey, awsSecretAccessKey, awsSessionToken, awsRegion},
		AssociatedBrowserScript:        nil,
		TaskFunctionOPSECPre:           nil,
		TaskFunctionParseArgString:     taskFunctionParseArgString,
		TaskFunctionParseArgDictionary: taskFunctionParseArgDictionary,
		TaskFunctionCreateTasking:      listCreateTask,
		TaskFunctionProcessResponse:    nil,
		TaskFunctionOPSECPost:          nil,
		TaskCompletionFunctions:        nil,
	}

	return command
}

func listCreateTask(task *structs.PTTaskMessageAllData) (resp structs.PTTaskCreateTaskingMessageResponse) {
	resp.TaskID = task.Task.ID

	provider, err := env.Get(task, "provider")
	if err != nil {
		err = fmt.Errorf("there was an error getting the 'provider' argument: %s", err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	var stdout string

	switch strings.ToLower(provider) {
	case "bedrock":
		bedrockModels, err := b.ListFoundationalModels(task)
		if err != nil {
			resp.Error = fmt.Sprintf("Failed to list Bedrock models: %s", err.Error())
			resp.Success = false
			logging.LogError(err, "there was an error listing bedrock models")
			return
		}

		for _, m := range bedrockModels {
			stdout += fmt.Sprintf("%s\n", m)
		}
	case "openai":
		stdout, err = openai.List(task)
		if err != nil {
			resp.Error = fmt.Sprintf("Failed to list OpenAI models: %s", err.Error())
			resp.Success = false
			logging.LogError(err, "there was an error listing OpenAI models")
			return
		}
	case "ollama":
		stdout, err = ollama.List(task)
		if err != nil {
			resp.Error = fmt.Sprintf("Failed to list ollama models: %s", err.Error())
			resp.Success = false
			logging.LogError(err, "there was an error listing ollama models")
			return
		}
	case "openwebui":
		stdout, err = openwebui.List(task)
		if err != nil {
			resp.Error = fmt.Sprintf("Failed to list OpenWebUI models: %s", err.Error())
			resp.Success = false
			logging.LogError(err, "there was an error listing openwebui models")
			return
		}
	}

	msg := mythicrpc.MythicRPCResponseCreateMessage{
		TaskID:   task.Task.ID,
		Response: []byte(stdout),
	}

	r, err := mythicrpc.SendMythicRPCResponseCreate(msg)
	if err != nil {
		resp.Error = fmt.Sprintf("Failed to send response: %s", err.Error())
		resp.Success = false
		logging.LogError(err, "there was an error sending the Mythic RPC Response Create message")
		return
	}

	resp.DisplayParams = &provider
	resp.Success = true
	resp.Completed = &r.Success
	return
}

File: Payload_Type/sage/container/agent_functions/list.py
from mythic_container.MythicCommandBase import TaskArguments, CommandBase, CommandParameter, ParameterType, ParameterGroupInfo, SupportedUIFeature, PTTaskMessageAllData, PTTaskCreateTaskingMessageResponse, PTTaskMessageTaskData
from mythic_container.MythicRPC import MythicRPCResponseCreateMessage, SendMythicRPCResponseCreate, MythicRPCCallbackUpdateMessage, SendMythicRPCCallbackUpdate, SendMythicRPCTaskUpdate, MythicRPCTaskUpdateMessage 
from mythic_container.logging import logger
from .utils import get_secret
from ai.model import Model, add_session, get_session, remove_session
import requests
from urllib.parse import urljoin

class ListArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)

        # Provider
        provider = CommandParameter(
            name="provider",
            display_name="Provider",
            cli_name="provider",
            type=ParameterType.String,
            description="The model provider to interact with (e.g. Anthropic, Bedrock, OpenAI)",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=3)]
        )

        # API Endpoint
        api_endpoint = CommandParameter(
            name="api_endpoint",
            display_name="API Endpoint",
            cli_name="api-endpoint",
            type=ParameterType.String,
            description="[OPTIONAL] The API endpoint to use for the selected provider",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=5)]
        )
        # API Key
        api_key = CommandParameter(
            name="api_key",
            display_name="API Key",
            cli_name="api-key",
            type=ParameterType.String,
            description="[OPTIONAL] The API key to use for the selected provider",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=6)]
        )
        # AWS Access Key ID
        aws_access_key = CommandParameter(
            name="AWS_ACCESS_KEY_ID",
            display_name="AWS_ACCESS_KEY_ID",
            cli_name="AWS_ACCESS_KEY_ID",
            type=ParameterType.String,
            description="[OPTIONAL] The AWS Access Key ID (AWS_ACCESS_KEY_ID) to use for Bedrock",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=7)]
        )
        # AWS Secret Access Key
        aws_secret_access_key = CommandParameter(
            name="AWS_SECRET_ACCESS_KEY",
            display_name="AWS_SECRET_ACCESS_KEY",
            cli_name="AWS_SECRET_ACCESS_KEY",
            type=ParameterType.String,
            description="[OPTIONAL] The AWS Secret Access Key (AWS_SECRET_ACCESS_KEY) to use for Bedrock",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=8)]
        )
        # AWS Session Token
        aws_session_token = CommandParameter(
            name="AWS_SESSION_TOKEN",
            display_name="AWS_SESSION_TOKEN",
            cli_name="AWS_SESSION_TOKEN",
            type=ParameterType.String,
            description="[OPTIONAL] The AWS Session Token (AWS_SESSION_TOKEN) to use for Bedrock",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=9)]
        )
        # AWS Region
        aws_region = CommandParameter(
            name="AWS_DEFAULT_REGION",
            display_name="AWS_DEFAULT_REGION",
            cli_name="AWS_DEFAULT_REGION",
            type=ParameterType.String,
            description="[OPTIONAL] The AWS Region (AWS_DEFAULT_REGION) to use for Bedrock",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=10)]
        )

        # Add all the parameters
        self.args = [provider, api_endpoint, api_key, aws_access_key, aws_secret_access_key, aws_session_token, aws_region]

    async def parse_arguments(self):
        pass

    async def parse_dictionary(self, dictionary_arguments):
        pass

class ListCommand(CommandBase):
    cmd = "list"
    needs_admin = False
    help_cmd = "list"
    description = "List all available models for the selected provider."
    version = 1
    author = "@Ne0nd0g"
    argument_class = ListArguments

    async def create_go_tasking(self, taskData: PTTaskMessageAllData) -> PTTaskCreateTaskingMessageResponse:
        response = PTTaskCreateTaskingMessageResponse(
            TaskID=taskData.Task.ID,
            Success=True,
        )

        provider = get_secret(taskData=taskData, key="provider")
        if provider is None:
            response.Success = False
            response.Error = "unable to find the 'provider' key in the task, user secrets, payload build parameters, or in the payload container's environment variables"
            return response
        
                # These can be None as they are optional
        api_endpoint = get_secret(taskData=taskData, key="API_ENDPOINT")
        api_key = get_secret(taskData=taskData, key="API_KEY")
        aws_access_key_id = get_secret(taskData=taskData, key="AWS_ACCESS_KEY_ID")
        aws_secret_access_key = get_secret(taskData=taskData, key="AWS_SECRET_ACCESS_KEY")
        aws_session_token = get_secret(taskData=taskData, key="AWS_SESSION_TOKEN")
        aws_region = get_secret(taskData=taskData, key="AWS_DEFAULT_REGION")

        if (provider.lower() == "openai" or provider.lower() == "anthropic") and not api_key:
            response.Error = "An API key was not found in the task, user secrets, payload build parameters, or in the payload container's environment variables. Please set the API key."
            response.Success = False
            return response

        response.DisplayParams = f'available models for {provider}'
 
        models_list : str = ""
        if provider.lower() == "openai":
            if not api_endpoint:
                # response.Error = "An API endpoint was not found in the task, user secrets, payload build parameters, or in the payload container's environment variables. Please set the API endpoint."
                api_endpoint = "https://api.openai.com"

            resp = requests.get(
                urljoin(api_endpoint, "v1/models"),
                headers={"Authorization": f"Bearer {api_key}"}
            )
            if resp.status_code != 200:
                response.Error = f"Failed to list models: {resp.text}"
                response.Success = False
                return response
            data = resp.json()
            for model in data["data"]:
                models_list += f"{model['id']}\n"
        elif provider.lower() == "anthropic":
            if not api_endpoint:
                api_endpoint = "https://api.anthropic.com"
            url = urljoin(api_endpoint, "v1/models")
            headers = {"x-api-key": f"{api_key}", "anthropic-version": "2023-06-01"}
            resp = requests.get(url, headers=headers)
            if resp.status_code != 200:
                response.Error = f"Failed to list models: {resp.text}"
                response.Success = False
                return response
            data = resp.json()
            for model in data["data"]:
                models_list += f"{model['id']}\n"
        elif provider.lower() == "bedrock":
            models_list = "Listing models for Bedrock is not supported yet. Please use the AWS CLI or SDK to list models."
        else:
            response.Error = f"Listing models for provider '{provider}' is not supported yet."
            response.Success = False
            return response
        
        resp = await SendMythicRPCResponseCreate(MythicRPCResponseCreateMessage(taskData.Task.ID, models_list.encode()))
        if not resp.Success:
            response.Success = False
            response.Error = resp.Error
        
        resp = await SendMythicRPCCallbackUpdate(
            MythicRPCCallbackUpdateMessage(
                TaskID=taskData.Task.ID, 
                UpdateLastCheckinTime=True,
                UpdateLastCheckinTimeViaC2Profile=""
            )
        )
        if not resp.Success:
            logger.error(f"Failed to update callback for task {taskData.Task.ID}: {resp.Error}")
        
        return response